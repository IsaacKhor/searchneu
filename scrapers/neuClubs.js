import request from 'superagent';
import * as acorn from 'acorn';
import cheerio from 'cheerio';
import fs from 'fs-promise';
import mkdirp from 'mkdirp-promise';
import path from 'path';
import macros from './macros';


// Example schema:
//   { audience: 'Undergraduate',
//     organizationemailaddress: '[redacted]@gmail.com',
//     facebookPage: 'https://www.facebook.com/[redacted]/',
//     meetingDay: 'Tuesday',
//     meetingTime: '7:00 PM',
//     meetingLocation: 'Lake Hall',
//     advisorName: 'Peter Simon',
//     presidentName: 'bob smith',
//     treasurerName: 'joe smith',
//     organizationMissionStatement: 'ECO.....',
//     name: 'EconPress',
//     site: 'http://www.northeastern.edu/econpress' },

// Not scraping right now: Description and some other stuff. No reason, just haven't had a reason to add it yet :P


async function scrapeDetails(url) {
  console.log(url);

  // Fire a get to that url
  const detailHtml = (await request.get(url)).text;

  // load the new html
  const $ = cheerio.load(detailHtml);

  // this part needs to be cleaned up but yea
  const detailElements = $('#profile_for_org > ul > li');

  const obj = {};

  // console.log(detailElements.length, url)
  // Looks like need to grab the href from the links but this is close
  for (let i = 0; i < detailElements.length; i++) {
    const $thisEle = $(detailElements[i]);

    // Get the key and convert it to cammel case
    let key = $('strong', $thisEle).text();

    key = key.trim();
    if (key.endsWith(':')) {
      key = key.slice(0, key.length - 1);
    }
    key = key.replace(/ /gi, '');
    key = key[0].toLowerCase() + key.slice(1);

    // Get the value of this field
    // If its a link to something, get all the links
    // If not, get the text
    let value;
    const possibleLinks = $('a', $thisEle);

    if (possibleLinks.length > 0) {
      const emails = [];
      for (let j = 0; j < possibleLinks.length; j++) {
        emails.push(possibleLinks[j].attribs.href);
      }
      value = emails;
    } else {
      value = $('p', $thisEle).text();
    }

    if (value) {
      obj[key] = value;
    }
  }

  // Remove mailto: from the beginning of the email addr
  if (obj.organizationemailaddress && obj.organizationemailaddress.length > 0) {
    for (let j = 0; j < obj.organizationemailaddress.length; j++) {
      const email = obj.organizationemailaddress[j];
      if (email.startsWith('mailto:')) {
        obj.organizationemailaddress[j] = email.slice('mailto:'.length);
      }
    }
  }

  // Slurp up the name
  const name = $('#full_profile > h2').text().trim();
  obj.name = name;

  // and get link jawn too
  const site = $('#org_extranet_url > a');
  if (site.length > 0) {
    obj.site = site[0].attribs.href;
  }

  if (Object.keys(obj).length < 3) {
    console.log('Error', url);
  }

  return obj;
}


async function scrapeLetter(letter) {
  const resp = await request.post(`http://neu.orgsync.com/search/get_orgs_by_letter/${letter.toUpperCase()}`);

  // Abstract Syntax Trees are the data structure that is used to parse programming languages (like javascript)
  // Like, the first step of running a programming language is to parse it into a AST
  // then preform a bunch of optimizations on it
  // and then compile it to machine code (if it is something like C)
  // or just run the AST directly (like python or js)
  // google for more info, feel free to log this directly
  const ast = acorn.parse(resp.text);

  // The reponse that we get back from that url is two lines of JS code
  // We are looking for the string arguments in the second line
  const html = ast.body[1].expression.arguments[0].value;

  // Dope, lets treat that as HTML and parse it
  const $ = cheerio.load(html);

  // Get all the a elements from it
  const elements = $('a');

  const orgs = [];
  const promises = [];

  // Look through all the below elements
  for (let i = elements.length - 1; i >= 0; i--) {
    if (!elements[i].attribs || !elements[i].attribs.href || elements[i].attribs.href === '#' || !elements[i].attribs.href.includes('show_profile')) {
      continue;
    }

    promises.push(scrapeDetails(elements[i].attribs.href).then((org) => {
      orgs.push(org);
    }));
  }


  await Promise.all(promises);

  return orgs;
}


async function main() {
  let orgs = [];

  const promises = [];

  macros.ALPHABET.split('').forEach((letter) => {
    promises.push(scrapeLetter(letter).then((someOrgs) => {
      console.log(letter, 'had', someOrgs.length, 'orgs now at ', orgs.length);
      orgs = orgs.concat(someOrgs);
    }));
  });

  await Promise.all(promises);

  const outputPath = path.join(macros.PUBLIC_DIR, 'getClubs', 'neu.edu');

  await mkdirp(outputPath);

  await fs.writeFile(path.join(outputPath, 'data.json'), JSON.stringify(orgs));
  console.log('done!', orgs.length);
}

if (require.main === module) {
  main();
}


export default main;