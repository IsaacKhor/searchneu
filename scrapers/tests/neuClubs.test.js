import fs from 'fs-promise'

import neuClubs from '../neuClubs';

fit('parseDetails should work', async function(done) {
	var data = await fs.readFile('scrapers/tests/data/neuClubs/club.json')
	var resp = JSON.parse(data)
	let output = neuClubs.parseDetails(resp.body)

	expect(output).toMatchSnapshot()
	done()
});
 
fit('parseLetterAndPage should work', async function(done) {
	var data = await fs.readFile('scrapers/tests/data/neuClubs/letter.json')
	var resp = JSON.parse(data)
	let output = neuClubs.parseLetterAndPage(resp)

	expect(output).toMatchSnapshot()
	done()
});
 

// it('parseDetailpage', async function(done) {
// 	var data = await fs.readFile('scrapers/tests/data/neuClubs/person.json')
// 	var resp = JSON.parse(data)

// 	let output = neuClubs.parseDetailpage(resp)

// 	expect(output).toMatchSnapshot()  
// 	done()
// });
 
