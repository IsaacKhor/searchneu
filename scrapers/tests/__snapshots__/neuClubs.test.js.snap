// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseDetails should work 1`] = `
Object {
  "advisorName": "Allison Cockshaw-Betsold",
  "audience": "Undergraduate",
  "category": "Performing Arts",
  "description": "The Northeastern University Pep Band plays at men's and women's hockey, basketball, soccer, and volleyball home games along with other special events on campus and in the surrounding community.",
  "facebookPage": Array [
    "https://www.facebook.com/pages/NU-Pep-Band/",
  ],
  "instagram": "instagram.com/gonupepband",
  "meetingDay": "Council Meetings: Alternating Sundays, Rehearsals: Selected Mondays",
  "meetingLocation": "Council Meetings: See Calendar, Rehearsals: Fenway Center",
  "meetingTime": "Council Meetings: 5:00p - 7:00p, Rehearsals: 8:30p - 10:00p",
  "name": "NU Pep Band",
  "officelocation": "146 CSC",
  "orgPortalLink": "https://orgsync.com/27725/chapter",
  "organizationMissionStatement": "The purpose of this organization shall be to serve the entire Northeastern University community through its support of the athletic teams and through its participation in non-athletic University and community events.",
  "organizationemailaddress": Array [
    "nupepband@gmail.com",
  ],
  "presidentName": "C. Justin Brown",
  "site": "http://www.northeastern.edu/pepband/",
  "treasurerName": "Amy Sevigny",
  "twitterHandle": Array [
    "https://twitter.com/gonupepband",
  ],
}
`;

exports[`parseLetterAndPage should work 1`] = `
Array [
  "http://neu.orgsync.com/show_profile/30932-student-activities-business-office-campus-resource",
  "http://neu.orgsync.com/show_profile/61469-student-academy-of-audiology",
  "http://neu.orgsync.com/show_profile/27977-strong-women-strong-girls",
  "http://neu.orgsync.com/show_profile/67151-stepping-on-another-level",
  "http://neu.orgsync.com/show_profile/127720-stemout",
  "http://neu.orgsync.com/show_profile/135212-spoon-university-northeastern",
  "http://neu.orgsync.com/show_profile/63104-spinning-arts-club",
  "http://neu.orgsync.com/show_profile/27997-spectrum-literary-arts-magazine",
  "http://neu.orgsync.com/show_profile/79680-somali-students-association",
  "http://neu.orgsync.com/show_profile/33253-sojourn-christian-collegiate-ministry-neu",
  "http://neu.orgsync.com/show_profile/27983-society-of-women-engineers",
  "http://neu.orgsync.com/show_profile/124721-society-of-physics-students",
  "http://neu.orgsync.com/show_profile/27959-society-of-hispanic-professionals-engineers",
  "http://neu.orgsync.com/show_profile/67847-society-of-asian-scientists-and-engineers",
  "http://neu.orgsync.com/show_profile/28034-social-enterprise-institute-student-association",
  "http://neu.orgsync.com/show_profile/30386-slow-food-nu",
  "http://neu.orgsync.com/show_profile/27817-sinha-capoeira",
  "http://neu.orgsync.com/show_profile/28799-silver-masque",
  "http://neu.orgsync.com/show_profile/27793-sigma-sigma-sigma",
  "http://neu.orgsync.com/show_profile/27960-sigma-phi-epsilon",
  "http://neu.orgsync.com/show_profile/71761-sigma-kappa",
  "http://neu.orgsync.com/show_profile/27853-sigma-delta-tau-gamma-mu",
  "http://neu.orgsync.com/show_profile/27727-sigma-beta-rho-fraternity-inc",
  "http://neu.orgsync.com/show_profile/145923-sexual-assault-response-coalition",
  "http://neu.orgsync.com/show_profile/145821-senior-legacy-committee-campus-resource",
  "http://neu.orgsync.com/show_profile/29267-senior-class-events-committee",
  "http://neu.orgsync.com/show_profile/127435-secular-humanist-society",
  "http://neu.orgsync.com/show_profile/97461-scientista",
  "http://neu.orgsync.com/show_profile/29390-saudi-arabian-student-organization",
  "http://neu.orgsync.com/show_profile/32710-sanskriti",
]
`;
